File: scmeta/scmeta.clj
#!/usr/bin/env bb -o

(require '[docopt.core :as docopt]
         '[babashka.process :refer [sh]]
         '[cheshire.core :as json]
         '[clojure.string :as str])

(defn wrap-as-time
  "Takes a given `time` in seconds and returns it as a time String.\n
   This is useful for working with dataframes. (appends an s)"
  [time]
  (str time "s"))

(defn parse-int
  "Takes a given String [s] and returns an Integer."
  [s]
  (Integer/parseInt (re-find #"\A-?\d+" s)))

(defn read-hf-file
  "Takes a hyperfine.json result file and returns a flattened hash-map."
  [file]
  (->  file
       (slurp)
       (json/parse-string true)
       :results
       first))

(defn read-pi-file
  "Takes a pi.txt file and returns the content as a trimmed String."
  [file]
  (-> file
      (slurp)
      (str/trim-newline)))

(defn pi-accuracy
  ;; https://github.com/niklas-heer/speed-comparison/issues/78#issuecomment-1292708468
  "Takes a given number for Pi and returns the accuracy.

   The higher the number the better:
     - pi-accuracy(3.14) = 3.29
     - pi-accuracy(3.145) = 2.96
     - pi-accuracy(3.1415) = 4.53
     - pi-accuracy(3.1415926535897) = 13.5"
  [pi]
  (let [x (Double/parseDouble pi)]
    ;; Formular: accuracy(x) = -log10(abs(1-x/pi))
    (- (Math/log10 (abs (- 1 (/ x Math/PI)))))))

(defn get-version
  "Takes a command as `cmd` which gets executed as a shell process.
   The command should return some version number like: `1.0.0`

   The second parameter `index` determines which version number to return, should there be multiple.

   Returns a version number as String."
  [cmd index]
  (-> cmd
      (sh)
      :out
      ;; re-seq needs the input as the last argument so we use "thread-last" (->>)
      (->> (re-seq #"\d+(\.\d+)+"))
      (nth index)
      first))

(defn generate-json [arg-map]
  (let [lvi       (parse-int    (arg-map "--lvi"))
        lv        (get-version  (arg-map "--lv") lvi)
        pi        (read-pi-file (arg-map "--pi"))
        hyperfine (read-hf-file (arg-map "--hf"))
        accuracy  (pi-accuracy pi)]
    (json/generate-string {:Language         (arg-map "--ln")
                           :Version          lv
                           :Command          (hyperfine :command)
                           :CalculatedPi     pi
                           :Accuracy         accuracy
                           :Mean             (wrap-as-time (hyperfine :mean))
                           :Median           (wrap-as-time (hyperfine :median))
                           :Min              (wrap-as-time (hyperfine :min))
                           :Max              (wrap-as-time (hyperfine :max))
                           :Stddev           (wrap-as-time (hyperfine :stddev))
                           :UserTime         (wrap-as-time (hyperfine :user))
                           :SystemTime       (wrap-as-time (hyperfine :system))
                           :TimesPerRun      (hyperfine :times)
                           :ExitCodesPerRun  (hyperfine :exit_codes)}
                          {:pretty true})))

(def usage "SCMeta.

Usage:
  scmeta [options]

Options:
  --ln=<name>   Name of the language.
  --lv=<cmd>    Command to get the language version.
  --lvi=<int>   If there are multiple versions for `lv`, select the index [default: 0].
  --hf=<json>   Hyperfine JSON file path.
  --pi=<txt>    Path to pi.txt file.
  --out=<json>  Path to write the output to")

(docopt/docopt usage
               *command-line-args*
               (fn [arg-map]
                 (->> arg-map
                      (generate-json)
                      (spit (arg-map "--out")))))

File: dagger-poc/scmeta.py
#!/usr/bin/env python3
"""
scmeta - Speed Comparison Metadata Generator

Processes hyperfine benchmark results and generates metadata JSON.
Compatible with MicroPython for minimal container footprint.

Usage:
    python scmeta.py --lang-name="Rust" --target-name="rust" \
        --lang-version="1.75.0" --hyperfine=hyperfine.json \
        --pi=pi.txt --output=result.json
"""

import json
import math
import sys

VERSION = "2.0.0"


def pi_accuracy(value_str):
    """Calculate accuracy of computed pi value.

    Returns the number of correct decimal places (higher is better).
    """
    try:
        value = float(value_str)
    except (ValueError, TypeError):
        raise ValueError(f"Invalid pi value: '{value_str}' is not a valid number")

    if value == 0.0:
        raise ValueError("Invalid pi value: cannot be zero")

    # math.pi is available in MicroPython
    accuracy = 1 - (value / math.pi)
    if abs(accuracy) == 0.0:
        return float("inf")

    return -math.log10(abs(accuracy))


def get_version(text, match_index=0):
    """Extract version number from text.

    Simple implementation without regex (MicroPython compatible).
    Finds patterns like "1.2.3" or "21.0.1".
    """
    versions = []
    i = 0
    while i < len(text):
        # Find start of a number
        if text[i].isdigit():
            start = i
            # Consume the version pattern: digits and dots
            while i < len(text) and (text[i].isdigit() or text[i] == "."):
                i += 1
            version = text[start:i]
            # Must contain at least one dot to be a version
            if "." in version and not version.endswith("."):
                versions.append(version)
        else:
            i += 1

    if not versions:
        raise ValueError(f"No version number found in: '{text}'")

    if match_index < 0 or match_index >= len(versions):
        available = "\n".join(f"  {i}: {v}" for i, v in enumerate(versions))
        raise ValueError(f"Invalid match index {match_index}. Available versions:\n{available}")

    return versions[match_index]


def to_timedelta(number):
    """Format number as seconds string."""
    return f"{number}s"


def parse_args(argv):
    """Parse command line arguments.

    Simple implementation without argparse (MicroPython compatible).
    """
    args = {
        "lang_name": None,
        "target_name": None,
        "lang_version": None,
        "hyperfine": None,
        "pi": None,
        "output": None,
    }

    i = 1  # Skip script name
    while i < len(argv):
        arg = argv[i]

        if arg in ("-v", "--version"):
            print(f"scmeta {VERSION}")
            sys.exit(0)

        if arg in ("-h", "--help"):
            print(__doc__)
            sys.exit(0)

        if "=" in arg:
            key, value = arg.split("=", 1)
        elif i + 1 < len(argv) and not argv[i + 1].startswith("-"):
            key = arg
            value = argv[i + 1]
            i += 1
        else:
            print(f"ERROR: {arg} requires a value", file=sys.stderr)
            sys.exit(1)

        # Map argument names to dict keys
        if key in ("--lang-name",):
            args["lang_name"] = value
        elif key in ("--target-name",):
            args["target_name"] = value
        elif key in ("--lang-version",):
            args["lang_version"] = value
        elif key in ("--hyperfine",):
            args["hyperfine"] = value
        elif key in ("--pi",):
            args["pi"] = value
        elif key in ("--output",):
            args["output"] = value
        else:
            print(f"ERROR: Unknown option {key}", file=sys.stderr)
            sys.exit(1)

        i += 1

    return args


def main():
    args = parse_args(sys.argv)

    # Validate required arguments
    required = ["lang_name", "target_name", "lang_version", "hyperfine", "pi", "output"]
    for name in required:
        if args[name] is None:
            opt_name = "--" + name.replace("_", "-")
            print(f"ERROR: {opt_name} is required!", file=sys.stderr)
            sys.exit(1)

    # Read hyperfine JSON
    try:
        with open(args["hyperfine"], "r") as f:
            hyperfine = json.load(f)
    except Exception as e:
        print(f"ERROR: Failed to read hyperfine file: {e}", file=sys.stderr)
        sys.exit(1)

    # Read computed pi value
    try:
        with open(args["pi"], "r") as f:
            computed_pi = f.read().strip()
    except Exception as e:
        print(f"ERROR: Failed to read pi file: {e}", file=sys.stderr)
        sys.exit(1)

    if not computed_pi:
        print("ERROR: Pi file is empty!", file=sys.stderr)
        sys.exit(1)

    # Calculate accuracy
    try:
        accuracy = pi_accuracy(computed_pi)
    except ValueError as e:
        print(f"ERROR: {e}", file=sys.stderr)
        sys.exit(1)

    # Extract version from provided version string
    try:
        version = get_version(args["lang_version"])
    except ValueError as e:
        # If no version pattern found, use the raw string
        version = args["lang_version"].strip()

    # Get hyperfine results
    results = hyperfine.get("results", [{}])[0]

    # Build metadata
    metadata = {
        "Language": args["lang_name"],
        "Target": args["target_name"],
        "Version": version,
        "Command": results.get("command", ""),
        "CalculatedPi": computed_pi,
        "Accuracy": accuracy,
        "Mean": to_timedelta(results.get("mean", 0)),
        "Stddev": to_timedelta(results.get("stddev", 0)),
        "UserTime": to_timedelta(results.get("user", 0)),
        "SystemTime": to_timedelta(results.get("system", 0)),
        "Median": to_timedelta(results.get("median", 0)),
        "Min": to_timedelta(results.get("min", 0)),
        "Max": to_timedelta(results.get("max", 0)),
        "TimesPerRun": results.get("times", []),
        "ExitCodesPerRun": results.get("exit_codes", []),
    }

    # Write output
    try:
        with open(args["output"], "w") as f:
            # MicroPython's json.dump doesn't have indent, so we format manually
            json_str = json.dumps(metadata)
            # Pretty print if possible (CPython), otherwise just dump
            try:
                f.write(json.dumps(metadata, indent=2))
            except TypeError:
                f.write(json_str)
    except Exception as e:
        print(f"ERROR: Failed to write output file: {e}", file=sys.stderr)
        sys.exit(1)

    print("Successfully created metadata")
    print(f"Language: {args['lang_name']} ({version})")
    print(f"Output: {args['output']}")


if __name__ == "__main__":
    main()
